### PR TITLE
plugin/prometheus: write rcode properly to the metrics

### LINKED ISSUE #5125: CoreDNS response rcode is incorrectly reported in prometheus metrics
**What happened**:
Hello, after bump to the latest 1.8.7 CoreDNS we noticed that CoreDNS prometheus metric related to DNS responses (`coredns_dns_responses_total`) shows wrong rcode (label `rcode`). Even when resolution ends in NXDOMAIN, the metric shows it as NOERROR.

**What you expected to happen**:
I would expect that `coredns_dns_responses_total` metric would report proper rcodes of responses

**How to reproduce it (as minimally and precisely as possible)**:
Run CoreDNS 1.8.7 with Corefile
```
.:53 {
  forward . 8.8.8.8
  prometheus
}
```

Do the DNS resolution via CoreDNS, note the answer rcode is NXDOMAIN
```
dig @127.0.0.1  nxdomain.cz

; <<>> DiG 9.17.17 <<>> @127.0.0.1 nxdomain.cz
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 45556
;; flags: qr rd ra ad; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;nxdomain.cz.			IN	A

;; AUTHORITY SECTION:
cz.			900	IN	SOA	a.ns.nic.cz. hostmaster.nic.cz. 1643190594 900 300 604800 900

;; Query time: 64 msec
;; SERVER: 127.0.0.1#53(127.0.0.1) (UDP)
;; WHEN: Wed Jan 26 11:13:48 CET 2022
;; MSG SIZE  rcvd: 106
```

go to http://localhost:9153/metrics in browser and observe that the request was reported as NOERROR
```
# HELP coredns_dns_responses_total Counter of response status codes.
# TYPE coredns_dns_responses_total counter
coredns_dns_responses_total{plugin="forward",rcode="NOERROR",server="dns://:53",zone="."} 1
```

**Anything else we need to know?**:
I think this was introduced as part of https://github.com/coredns/coredns/pull/4914 as the Recorder specific to metrics plugin does not write Rcode field properly https://github.com/coredns/coredns/pull/4914/files#diff-17e896bba0ee296638a8e2cf4cecd323b1a14a9e0a6c6bbb6f54bf88575d4b71R23

**Environment**:
linux

- the version of CoreDNS: 1.8.7
- Corefile:
```
.:53 {
  forward . 8.8.8.8
  prometheus
}
```
- logs, if applicable:
- OS (e.g: `cat /etc/os-release`):
- Others:

### PR DESCRIPTION
**Pull request by Tantalor93**
Signed-off-by: Ondřej Benkovský <ondrej.benkovsky@jamf.com>

### 1. Why is this pull request needed and what does it do?
This PR adjust recorder in prometheus plugin in a way that it properly propagated rcode written from following plugins

### 2. Which issues (if any) are related?
close #5125 

### 3. Which documentation changes (if any) need to be made?
none

### 4. Does this introduce a backward incompatible change or deprecation?
no

### CODE REVIEW

File: plugin/metrics/recorder.go [hunk 1]
```diff
@@ -24,7 +24,5 @@ func (r *Recorder) WriteMsg(res *dns.Msg) error {
 	_, r.Caller[0], _, _ = runtime.Caller(1)
 	_, r.Caller[1], _, _ = runtime.Caller(2)
 	_, r.Caller[2], _, _ = runtime.Caller(3)
-	r.Len += res.Len()
-	r.Msg = res
-	return r.ResponseWriter.WriteMsg(res)
+	return r.Recorder.WriteMsg(res)
```
Comment (Tantalor93) [inline plugin/metrics/recorder.go:27]: I am using embeded `dnstest.Recorder` to properly set `r.Len`, `r.Msg`, `r.Rcode` fields

**Review by greenpau — CHANGES_REQUESTED** [review]
@Tantalor93 , makes sense! 👍 Please add a test for this.

File: plugin/metrics/recorder_test.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			tw := inmemoryWriter{ResponseWriter: test.ResponseWriter{}}
+			rec := NewRecorder(&tw)
+
+			if err := rec.WriteMsg(tt.msg); err != nil {
+				t.Error("WriteMsg() unexpected error", err)
+			}
+
+			if rec.Msg != tt.msg {
```
Comment (greenpau) [inline plugin/metrics/recorder_test.go:57]: @Tantalor93 , please the following pattern when outputting errors:

https://github.com/coredns/coredns/blob/2009df1474124ddaf6d6955e0864a3ef4f0feb14/plugin/metrics/metrics_test.go#L78

**Review by greenpau — CHANGES_REQUESTED** [review]

**Review by greenpau — APPROVED** [review]
@Tantalor93 , LGTM 👍 Thank you for contributing!

**Review by chrisohaver — APPROVED** [review]

**Pull request merged**